### PR TITLE
Fetch and Publish System Status Info

### PR DESCRIPTION
**Pull request by harlem88**
- Publish System Status Info to remote Astarte instance using io.edgehog.devicemanager.SystemStatus interface.
- Add edgehog_device_handle to interact with edgehog_device api.

### CODE REVIEW

File: src/edgehog.c [hunk 1]
```diff
@@ -17,5 +17,167 @@
...
+    if (ret != ASTARTE_OK) {
+        ESP_LOGE(TAG, "Unable to add Astarte Interface ( %s ) error code: %d",
+            hardware_info_interface.name, ret);
+        return ESP_FAIL;
+    }
+
+    ret = astarte_device_add_interface(device, &os_runtime_status_interface);
+    if (ret != ASTARTE_OK) {
+        ESP_LOGE(TAG, "Unable to add Astarte Interface ( %s ) error code: %d",
+            hardware_info_interface.name, ret);
```
Comment (FrancescoVaiani) [inline src/edgehog.c:68]: qui dovrebbe essere `os_runtime_status_interface.name` e non `hardware_info_interface.name`

File: src/edgehog.c [hunk 2]
```diff
@@ -87,20 +141,44 @@ void publish_device_hardware_info(astarte_device_handle_t device)
...
 }
 
-void edgehog_init(astarte_device_handle_t astarte_device)
+void publish_system_status(edgehog_device_handle_t edgehog_device)
 {
-    ESP_ERROR_CHECK(add_interfaces(astarte_device));
-    publish_device_hardware_info(astarte_device);
+    int64_t uptime_millis = esp_timer_get_time() / 1000;
+    uint32_t avail_memory = esp_get_minimum_free_heap_size();
+    int task_count = (int) uxTaskGetNumberOfTasks();
```
Comment (bettio) [inline src/edgehog.c:160]: I don't understand why you are doing this cast.

File: src/edgehog.c [hunk 3]
```diff
@@ -87,20 +141,44 @@ void publish_device_hardware_info(astarte_device_handle_t device)
...
 }
+
+void edgehog_device_destroy(edgehog_device_handle_t edgehog_device)
+{
+    if (edgehog_device) {
+        astarte_device_destroy(edgehog_device->astarte_device);
+    }
+
+    free(edgehog_device);
+}
```
Comment (bettio) [inline src/edgehog.c:184]: missing new line at the end of file.

File: src/edgehog.c [hunk 4]
```diff
@@ -18,19 +18,65 @@
...
+          .major_version = 0,
+          .minor_version = 1,
+          .ownership = OWNERSHIP_DEVICE,
+          .type = TYPE_DATASTREAM };
+
+esp_err_t add_interfaces(astarte_device_handle_t astarte_device);
+void publish_device_hardware_info(astarte_device_handle_t astarte_device);
+void publish_system_status(edgehog_device_handle_t edgehog_device);
+
+edgehog_device_handle_t edgehog_init(astarte_device_handle_t astarte_device)
```
Comment (bettio) [inline src/edgehog.c:56]: let's call it `edgehog_new` since we are allocating memory, otherwise `edgehog_init` if we initialize already allocated memory.

**Review by bettio — CHANGES_REQUESTED** [review]

File: src/edgehog.c [hunk 5]
```diff
@@ -87,20 +141,44 @@ void publish_device_hardware_info(astarte_device_handle_t device)
...
+    astarte_device_stream_string(
+        astarte_device, hardware_info_interface.name, "/cpu/architecture", cpu_architecture, 2);
+    astarte_device_stream_string(
+        astarte_device, hardware_info_interface.name, "/cpu/model", cpu_model, 2);
+    astarte_device_stream_string(
+        astarte_device, hardware_info_interface.name, "/cpu/modelName", cpu_model_name, 2);
+    astarte_device_stream_string(
+        astarte_device, hardware_info_interface.name, "/cpu/vendor", cpu_vendor, 2);
+    astarte_device_stream_integer(
+        astarte_device, hardware_info_interface.name, "/mem/totalBytes", mem_total_bytes, 2);
```
Comment (FrancescoVaiani) [inline src/edgehog.c:153]: I think we have a little regression here. This lines should be `astarte_device_set_[type]_property` as for previous commit.

**Review by FrancescoVaiani — CHANGES_REQUESTED** [review]

File: src/edgehog.c [hunk 6]
```diff
@@ -18,19 +18,65 @@
...
+const static astarte_interface_t system_status_status_interface
+    = { .name = "io.edgehog.devicemanager.SystemStatus",
+          .major_version = 0,
+          .minor_version = 1,
+          .ownership = OWNERSHIP_DEVICE,
+          .type = TYPE_DATASTREAM };
+
+esp_err_t add_interfaces(astarte_device_handle_t astarte_device);
+void publish_device_hardware_info(astarte_device_handle_t astarte_device);
+void publish_system_status(edgehog_device_handle_t edgehog_device);
```
Comment (bettio) [inline src/edgehog.c:54]: these functions are not exposed in the .h file, so I suppose they are private. In that case I would mark them as static so their symbol in not exported outside the object file.

File: include/edgehog.h [hunk 7]
```diff
@@ -19,13 +19,30 @@
...
+
 #ifdef __cplusplus
 extern "C" {
 #endif
 
 #include <astarte_device.h>
+/**
+ * @brief create Edgehog device handle.
+ *
+ * @details This function has to be called to create the device SDK before doing anything else.
```
Comment (bettio) [inline include/edgehog.h:32]: I think we might rephrase it in a better way.

File: src/edgehog.c [hunk 8]
```diff
@@ -88,19 +142,43 @@ void publish_device_hardware_info(astarte_device_handle_t device)
...
+        astarte_device, hardware_info_interface.name, "/mem/totalBytes", mem_total_bytes);
 }
 
-void edgehog_init(astarte_device_handle_t astarte_device)
+void publish_system_status(edgehog_device_handle_t edgehog_device)
 {
-    ESP_ERROR_CHECK(add_interfaces(astarte_device));
-    publish_device_hardware_info(astarte_device);
+    int64_t uptime_millis = esp_timer_get_time() / 1000;
+    uint32_t avail_memory = esp_get_minimum_free_heap_size();
```
Comment (bettio) [inline src/edgehog.c:159]: I think here we should call `esp_get_free_heap_size()`, `esp_get_minimum_free_heap_size()` looks like a monotonic minimum. So let's check the documentation twice.

**Review by bettio — APPROVED** [review]

**Review by bettio — CHANGES_REQUESTED** [review]
there are some minor issues that need to be fixed.

**Review by bettio — APPROVED** [review]

**Review by FrancescoVaiani — APPROVED** [review]

**Pull request merged**